### ISSUE TITLE
New block ID mapping in the upcoming MC versions (introduced with 17w48a)

### ISSUE THREAD

**Issue by broese** [broese/mcbuild#20]
17w48a introduces a completely revamped system for block IDs, eliminating the old blockID/meta system.
A completely new implementation is needed for most subsystems in MCB.

**Comment by harbaughfan** [comment]
Hi, I would like to help in what little way I can.  Could you create a development branch and I'll issue a pull request, please?  
I forked and inserted a new protocol 404, updated packet ids etc, and figured out which pids still work and commented out the broken ones. 
I love your application & I'm so sad without it!  We can do it if we tackle in stages!

**Comment by broese** [comment]
Hi! Thanks for the offer! I must apologize again, I had I few turbulences in my life recently and neglected the project completely. I started implementing a few new things for the new framework / version 3.0 quite a while ago, but then couldn't continue.

How do you have it in mind? We just introduce a "dev_3.0" branch and start updating it with your/mine contributions? The "master" branch defines the latest rolling release - i.e. this is where MCB installations auto-update. I can't merge anything to there until a proper v3.0 is ready - or the installed clients will break.

The main issue I see are not so much the block IDs, but their states - rotation/etc. This system has changed completely. The building routines had many things hardcoded and I really need to come up with a new mechanism to handle this data and if possible provide compatibility with pre-1.13 protocol too.

Give me some time, I'll review what I had done before and push the dev branch probably tomorrow.

**Comment by harbaughfan** [comment]
Sounds great, I'll sit tight and watch for the initial dev branch.  In the mean time I will read up on the flattening.  
I was thinking an incremental approach would be ok (starting with packet decoding, new slot/item IDs, things that don't rely on block states like autoshear).  Then design a new build plan structure & get it building a floor.  Then rewrite the advanced build routines one-by-one.  I can totally see scrapping backwards compatibility since the new system is so different, and there will always be v2.0 for those who are still playing 1.12.

Thank you so much! It's a magnificent application.

**Comment by harbaughfan** [comment]
As for how the contributing will work, I think leader/helper is best.  Example you design the new struct that represents a block, and I can help with getting every block’s properties into the struct you design.  Just an example, but let’s start with this, if you find yourself faced with any drudgery, then assign it to me so you can get back to masterminding.  Let’s get in contact via email/im and I’m at your service.

**Comment by broese** [comment]
Hi again. I reviewed what I did in August/September, now I have a better understanding what I planned back then. I pushed the dev_3.0 branch to github

One thing I wanted to do was to add a proper event loop and I already wrote a little library for that.
At the moment, packets are handled synchronously (process packet, send and forget), and all the difficult actions like handling inventory, eating, etc. are more or less hacks. I just send out a bunch of packets hoping they will do whatever needed but sometimes it messes things up. I wanted to add properly timed asynchronous actions. This I think can be postponed for now, so I moved everything related to a separate branch. But it would be really nice to implement this in the future.

Now for the 1.13 support. I did following:
- made an empty SUPPORT_1.13 table - that was enough for the basic proxy to work and also to create protocol dumps (saved/*.mcs files)
- haven't looked into 1.13.1 and 1.13.2 protocols but they seem pretty much identical at a glance 
disabled gs_packet and gm_packet - these functions process packets for the gamestate and for the game actions
  The gamestate acts passively and just collects useful data to maintain the client's view of the situation - block data, entities, inventory state etc.
  The game module acts actively and decides how to react on incoming packets and uses gamestate as database
  Both of them started segfaulting when I tried to pass some packets to them because of unexpected block IDs.
- started enabling and implementing one by one some of the simple packets and dump them to see how they decode

I thought a bit how we can plan this. There are a couple of areas to work on, and unfortunately a lot needs to be done in mcp_packet, mcp_gamestate, mcp_game and mcp_ids and some helper modules before anything build-related will work. I guess we could start with the packets and check through all of them, one by one. Enable them in the table, check the format with the wiki, check the dump output, check if 1.13.x are compatible, commit, rinse, repeat. Not all packets are needed for building support - everything related to entities is more for the gameplay functions like autokill, etc. so they can be postponed. What we need is:

Player:
    SP_PlayerPositionLook
    CP_Player
    CP_PlayerPosition
    CP_PlayerLook
    CP_PlayerPositionLook
    SP_JoinGame
    SP_Respawn
    SP_ChangeGameState
    CP_TeleportConfirm
    CP_UseEntity
    CP_Player
    CP_PlayerPosition
    CP_PlayerPositionLook
    CP_PlayerLook
    CP_Animation
    CP_UseItem
    
Blocks:
    SP_BlockChange
    SP_MultiBlockChange
    SP_Explosion
    SP_UnloadChunk
    SP_ChunkData
    CP_PlayerBlockPlacement

Chat:
    SP_ChatMessage
    CP_ChatMessage

Inventory:
    SP_ConfirmTransaction
    SP_CloseWindow
    SP_OpenWindow
    SP_WindowItems
    SP_SetSlot
    SP_HeldItemChange
    CP_ConfirmTransaction
    CP_ClickWindow
    CP_CloseWindow
    CP_HeldItemChange
    CP_PickItem (maybe? this one is new)

I already had some of the blocks-related packets implemented. If you want to participate, you can take over some in the list above or check if the versions 1.13/1.13.1/1.13.2 are all identical for these packets.

Special care needed for the changing data formats - bid_t, slot_t have changed, maybe something else. It would be nice if you could check for those as you go through the packets so we won't miss anything. That is just the beginning though. Next step will be to reimplement the gamestate - player, block and inventory state.

Regarding the block IDs - there is a possibility to extract this data in JSON format from the official server. I'll write a script to parse it and convert it to C code, so I won't need to write it all by hand. But then I need to decide on internal format for this table.

**Comment by harbaughfan** [comment]
Sounds great! I’ll give each packet id a rigorous test tonite and let you
know.  I did just that over the weekend but not rigorously and only for
13.2

**Comment by broese** [comment]
I just checked the 1.13.1 and 1.13.2 support and everything seemed to work with the same table, for these few packets at least.
I think I'll look into SP_ChunkData tomorrow

**Comment by harbaughfan** [comment]
Hi, I confirmed many of these and pushed the changes to my fork, which you can pull.

Here are the ones with an issue:
SP_ChunkData
SP_WindowItems (broken because of a small change to Slot struct in 13.2)
SP_SetSlot (broken because of a small change to Slot struct in 13.2)
CP_ConfirmTransaction (was not supported in the previous version)
CP_ClickWindow (broken because of a small change to Slot struct in 13.2)
CP_PickItem (didn't check yet since it's new)

I believe I implemented all the rest in your list. For these I confirmed structs on wiki & got the packet to fire & dump in-game.

Small aside for SP_ChatMessage, it works and dumps the json but I'm not sure if the deocde_chat_json is still working... I couldn't get that part to fire.

**Comment by broese** [comment]
Thanks! I just merged your changes.
I think SP_ChatMessage was never properly decoded from JSON before, because I either passed it through or generated my own messages. The function only Q&D-decodes messages from chat so they appear as a clean string in the dump. Quite possible it's broken now because of format changes in JSON. Good thing they didn't change the CP_ChatMessage to JSON.

I tested SP_ChunkData briefly, the decoder from 1.9.4 seems to work, but I have to verify what kind of block IDs I get from it. I'll proceed with fixing the slot decode, then try reactivating the gamestate so I can check if blocks are stored correctly.

**Comment by harbaughfan** [comment]
After work tonite I'll go through and test/confirm more of the packet ids not on the list but I will withhold making any changes.  Then once you're done for the night, push your changes & I'll grab your new packet.c and commit/push for each of the packet id's I test.  I'll stay away from block/slot packets for tonight... just grabbing all the low hanging fruit that hasn't changed.

**Comment by broese** [comment]
Sorry, took me longer than expected.
I found out the 1.13/1.13.1 protocols were using a different slot format than 1.13.2.
I decided to drop support for them and limit everything to 1.13.2, because it's just not worth the trouble.
I implemented all the packets remaining in the list and also reenabled SP_ChunkData which seems to work with the 1.9.4 decoder. But that still needs to be verified in detail.
Turns out, there are differences in block/item IDs between the versions (facepalm). Which means we cannot define them statically, but need to load the corresponding JSON extracted from the server every time when connecting. Maybe it's better that way, I don't know.
And then there is even a good news - that new CP_PickItem packet might turn very useful because it could save some trouble picking materials from the inventory to the hotbar. Now the server can basically do that for you.

**Comment by harbaughfan** [comment]
Great news that's a ton of progress so fast! Tonight I grabbed your new packet.c & added support for 6 packet id's that I tested/confirmed.  I have 12 more to test.  If a packet has issues I am leaving it disabled & after I'm done I'll write up the pids with issues.  The issues are all irrelevant at the moment, so let's not slow down for them yet. If anything relevant comes up I'll let you know right away.  Thank you!

**Comment by harbaughfan** [comment]
Update: I tested the remaining 12 and activated the packet ids that are confirmed to work.  There are issues with 6 packet ids which I left unactivated.

Issues (Unactivated):
0x03,SP_SpawnMob,   Not sure what the problem is yet
0x04,SP_SpawnPainting,   (Struct change... Pstr(title) should be changed to Pvarint(motive)
0x05,SP_SpawnPlayer, (Unknown status -- I can't get this packet to fire)
0x0a,SP_BlockAction,  (BLOCKTYPE() is decoding the wrong blockname)
0x26,SP_Map,  (Struct change)
0x3f,SP_EntityMetadata,   (read_metadata fails on  assert(key < 32);)

**Comment by harbaughfan** [comment]
Hi, I'm at your service all weekend if there's anything you identify that I can take off your plate as a side project.  I get the sense we are in 1-man project mode until you get the basic structures the way you want. But hoping I can help since I have a ton of time this weekend.  If you don't have anything offhand for me, I'll try to see if I can find anything myself.  One idea I had while testing the packets was to get all the int-enums to display the meaning in text when dumping the packet.  You already do that in a few of the packets. And it would be something simple & out of the way, maybe not to useful but may help debugging in the future.  I can also tackle the packets with issues, but I was thinking struct changes are better left to you because you'll know better what it affects & how you want them, if there's any hope of reincorporating the other protocols someday etc.  But if you want me to push struct changes that's great too.  Just ideas in case there's nothing for me.

**Comment by broese** [comment]
Hi, thanks for that. Sorry if the coordination wasn't the best. I guess I can get some more work done on the weekend too.
Well, there is an idea for a bit of work - if it suits you. We will need a new module to load the block/item information from JSON data. This data can be exported from the server's jar file of the appropriate version using
`java -cp server.jar net.minecraft.data.Main --all`
That's a new function that was added with 1.13 release.

In generated/reports/ you then can find files blocks.json and items.json. We need to parse these to get IDs used in this particular protocol version. If you'd like, you could write a parser for these things using the json-c library that MCB already uses to parse the user profiles. Then some routines to lookup data by IDs and vv., so sort of a database that I could access to map IDs

Also some code to load/delete the databases. Since blocks database is huge, it also would make sense to maybe keep it in memory and only delete if a new protocol version was requested - so that the file does not need to be reparsed every time the user reconnects. My idea : define a "database_t" object that I can get from the load function and if that database was already once loaded, I just get the same copy again, without parsing the file again. Let's assume the json files are in database/ subdirectory and are identifable by their protocol ID, e.g. database/blocks_404.json and database/items_404.json

```
const database_t *load_database(int protocol_id);
unload_all_databases();
```
Then another issue: I'm not sure if this is legal to distribute JSON files from Mojang's data, and it would be too much to ask an average user to do it themselves. Maybe a better way would be an external tool to parse json and dump a database file in out own format - maybe even purely binary. This binary file is then distributed together with MCB - and is loaded quicker.

The data format for items could be pretty simple:
```
struct item_t {
  const char *name;
  int id;
};

/* example
"minecraft:obsidian": {
  "protocol_id": 139
}, => { "obsidian", 139 }  i.e. strip "minecraft:" part
*/

// lookup
int get_item_id(database_t *db, const char *name);
const char *get_item_name(database_t *db, int item_id);
get_item_id(db, "obsidian") => 139
get_item_name(db, 139) => "obsidian"
```

regarding the blocks, I need think of good way to organize them
cont…

**Comment by harbaughfan** [comment]
Sounds like a great project for me, I think step-by-step would work ok here.  I'll start by using the json library to parse the raw json file.  I'm thinking a database is an array of item structs with maybe some extra fields for protocol # and housekeeping.  I'll oversize the array and if I can't figure it out myself I'll lean on you to get the libhelper to properly size the array.  Then we move on to blocks, by then you'll know what struct you want.  After it's all working we can look at optimizing the database binary format on disk.  I know google has protobufs which is perfect for this kind of thing but I think it's great to avoid extra libraries if we can.

**Comment by broese** [comment]
What we need for a specific protocol are blocks and items, so I thought database_t would represent both for a specific protocol. When the mitm structure is initialized upon connection, I would query your module and get this database object and use it all the time.

Regarding the libraries - yes, if possible I would like to avoid using anything extra. They need to be available with Cygwin and common Linux distros. Still it would be PITA to update the package just to add a library.

Regarding the blocks data - it is more complicated, and I'm still figuring out how best deal with properties of the blocks. There are some needed for correct placement, some for switching and some are more or less irrelevant. I think there could be a static table mapping block symbolic names to a set of properties relevant to it. But that can wait for a while.

```
  "minecraft:oak_stairs": {
    "properties": {
      "facing": [
        "north",
        "south",
        "west",
        "east"
      ],
      "half": [
        "top",
        "bottom"
      ],
      "shape": [
        "straight",
        "inner_left",
        "inner_right",
        "outer_left",
        "outer_right"
      ],
      "waterlogged": [
        "true",
        "false"
      ]
    },
    "states": [
      {
        "properties": {
          "facing": "north",
          "half": "top",
          "shape": "straight",
          "waterlogged": "true"
        },
        "id": 1649
      },
      {
        "properties": {
          "facing": "north",
          "half": "top",
          "shape": "straight",
          "waterlogged": "false"
        },
        "id": 1650
      },
     ...
      {
        "properties": {
          "facing": "south",
          "half": "bottom",
          "shape": "outer_left",
          "waterlogged": "false"
        },
        "id": 1686
      },
     
const char * get_block_name(database_t *db, int id);
// (db,14) => "cobblestone"
// (db,1650) => "oak_stairs"
// (db,1686) => "oak_stairs"
const char * get_block_propval(database_t *db, int id, const char *propname);
// (db,14,"facing") => NULL // no such property
// (db,1650,"facing") => "north"
// (db,1686,"half") => "bottom"
struct prop_t { const char *pname, const char *pvalue };
// prop_t * defines a set of properties I'm interested in. It's a list no longer than 3 I guess
// how should we define the length? it could be {NULL,NULL} terminated or length explicitly given.
// many different IDs can match
int get_matching_block_ids(database_t *db, const char *name, prop_t *match, int *ids);
// places all ids matching a set of propeties for the block name into array ids (can be assumed to be long enough) and returns the number of ids
```

**Comment by harbaughfan** [comment]
Sounds great I'm diving in!

**Comment by broese** [comment]
I've reenabled gamestate and game modules, fixed a few things and now most gameplay-related things work again! Best thing - the inventory handling seems to be functional, at least it was working consistently for the things I tried. It's was pretty difficult to get it right back then and I'm happy they didn't mess it up.

If you have time, can you try testing it a bit, whether you find any inconsistencies. Just move the items around in the inventory and see if the output of #inv command matches what you see in the client. One thing that will not work for sure is the auto-transfer of armor items into armor slots.

**Comment by harbaughfan** [comment]
Sounds good I'll dive into testing inventory now & post back here.

**Comment by harbaughfan** [comment]
Same issue with auto-transfer of armor.. if I have a stack of something in my hotbar, then shift-click to move it into my inventory, #inv notices it gone from my hotbar but doesn't see it in the inventory.  Same issue if I have a stack of something in inventory & shift-click to move to hotbar.

PS an interesting comment at https://wiki.vg/How_to_Write_a_Client 
_The client has to know all clicking behaviour (including crafting recipes) and change the slots accordingly after each click, because the server does not send any Set Slot, just Confirm Transaction. But you can cheat and always send the wrong "Clicked item" in Click Window to force the server to send you all slots again while still applying your click._

Not sure your taste for an ugly hack like that... but if client's inventory is unsynched from gamestate in mcbuild, maybe a temporary route.

**Comment by broese** [comment]
Ah, what is NOT an ugly hack in this client? :)
Thanks for this info, I actually missed it before. If a window update can be somehow triggered, that would be actually awesome. I had some major issues with transactions failing in the past and could only get them somewhat reliably working after a huge amount of testing. If the inventory corruption is detected, MCB will actually ask the player to open any kind of container, so the inventory will be reset.
If the trick described in the article works, this could be used to reset the inventory without user interaction.
At the moment I actually have to implement all possible transactions to maintain the correct state. But at least, I do not have to deal with containers and crafting recipes, although I must check for the in-inventory crafting. And it is still probably buggy.

**Comment by harbaughfan** [comment]
PS if you are doing inventory testing, I uploaded code to display the item name after issuing #inv .  It's in the pull request.  Not sure if I am even on the right track with how I am setting up the db at all?  Does it go in separate c/h file?  Do you load_db each time you need a pointer to the struct no matter which file the code is in?  Stuff like that is a bit over my head.

**Comment by broese** [comment]
Typically, I would put the types definitions and the function prototypes into .h file and the implementation and anything that is to be hidden from other modules into .c file of the same name. So, how you made mcp_database.h and mcp_database.c you are totally correct. Just maybe update the copyright - if you are writing the module, you should put your name on it.

My idea was to call load_database just once upon new connection, get the pointer to the object, store it in the mitm struct and use it as a handle when I query it.

I reviewed your updates and have a few comments:
- you are returning database_t struct directly from the function - that's a bad idea for a large struct like this - you have 2 large arrays inside. Better would be to malloc it, initialize, and return a pointer.
- If you want to use the lh_arr.h fucntions from libhelper: this is how I would organize it:

```
typedef struct database_t {
  ...
  lh_arr_declare(item_t, item);  // declare a resizable array "item" with elements of type item_t
  lh_arr_declare(block_t, block);
}

// What it expands to:
typedef struct database_t {
  ...
  item_t * item_p;
  ssize_t item_c;
  block_t * block_p;
  ssize_t block_c;
}
```

So, if you have a _database_t *db = load_database(...)_, you can access these fields like this :
```
P(db->item) => db->item_p // pointer to the array
C(db->item) => db->item_c // count of elements
```
But if you have is statically like : _database_t db;_ , you can access them as:
```
P(db.item) => db.item_p // pointer to the array
C(db.item) => db.item_c // count of elements
```

If you want to add another element to the array:
`item_t *it = lh_arr_new_c(GAR(db->item));`
It will add a new element item_t, resize the array as necessary, initialize the element with zeros (due to _c macro) and return a pointer to the new element. The macro _GAR(db->item)_ resolves to _db->item_p,db->item_c,256_ that serves as arguments for reallocation. Using _it_ as a pointer to the newly allocated element, you can initialize it as you wish

It's probably a bad style, but it worked for me and saved me a lot of boilerplate code. That was my primary reason to make libhelper back in the days

**Comment by harbaughfan** [comment]
Just a question on mitm structure, if I add a database pointer to it, and load database as soon as proxy_pump establishes a connection and determines protocolid, will the mitm database handle be a global variable so the it can be accessed from code inside another file like mcb_ids.c ?   I don't understand if that works/how to get that working.

**Comment by broese** [comment]
_mitm_ is a global variable - basically a global state of the proxy containing all the important stuff about the session, connections, encryption, etc.. OK, now I realize my mistake - _mitm_ is actually only used within mcproxy.c. The other modules are only getting specific MCPacket* and MCPacketQueue* references.

The other global variable used everywhere is _currentProtocol_, which is located in mcp_packet.c and is exported by _mcp_packet.h_ to other modules. It is initialized from mcproxy.c upon receiving CI_Handshake packet and calling setProtocol in mcp_packet.c

Maybe the idea for now would be to:
- have _database_t *db_ defined as a global variable in mcp_packet.c , just as _currentProtocol_
- initialize it in setProtocol, by calling load_database and storing it
- making it accessible to other modules via mcp_packet.h the same way as currentProtocol

Yeah, all this is quite a mess. My idea for a long time was to make things more object-oriented and get rid of all these global variables. Maybe it will be possible with the move to the proper event loop.

**Comment by harbaughfan** [comment]
Perfect thank you!  And I'm making progress on decoding blocks.json, I just got the blockname<->id storing in the database and querying name from id.  Next is block properties.

**Comment by harbaughfan** [comment]
Just a note on the block database, I am adding a function 
`int get_block_default_id(database_t *db, int id)`
Which given any blockid you encounter in the game, it will return its "default" id for that blockname.  so if you have 
```
  "minecraft:grass_block": {
    "properties": {
      "snowy": [
        "true",
        "false"
      ]
    },
    "states": [
      {
        "properties": {
          "snowy": "true"
        },
        "id": 8
      },
      {
        "properties": {
          "snowy": "false"
        },
        "id": 9,
        "default": true
      }
    ]
  },
```

If we don't care whether the grass block is snowy, we can just translate blockid 8 to blockid 9 by calling 
`get_block_default_id(8);  //value of 9`

**Comment by harbaughfan** [comment]
Hi, Probably need help at this point.  The code is parsing json like a champ, but I get lost with all the memory allocation, passing pointers between files etc. Code compiles & works though i get warnings, likely because of how the db variable/pointers are stored and passed.

Note on decoding block names:  0x10 SP_BlockAction still uses old style id's.  So when dumping it would need a seperate function to decode its old-style id.  I added old id's to the database and a routine "get_block_name_from_old_id"  but I think it's time for your eyes on how everything's setup first since it would take some circumvention for me to expose that method to the BlockAction packet-dump.

I couldn't figure out how to allocate memory for the db dynamically.. I was getting segfaults left and right.

**Comment by harbaughfan** [comment]
Hi, as far as the copyright notice, I transfer any possible copyright claims to you for everything I do.  You're the sole author, owner of this code, and copyright holder.  It's your application, I just want to help you in what little way I can.  Thank you for mentioning it.

**Comment by harbaughfan** [comment]
Hi, good news, I think I'm back on track with setting up the global variable the way you described.  One item I'd like to raise:
To avoid compiler warnings I had to change 
`const database_t *load_database(int protocol_id) {`
to 
`database_t *load_database(int protocol_id) {`
does that make sense or does that mean I am not grasping something about how the variables should be setup?

**Comment by harbaughfan** [comment]
Hi Just an update regarding distributing the blocks.json & items.json..  If load_database sees those files are missing, it will automatically download the correct server.jar and extract the files from it.   That way we don't have to distribute them.  I'm also working on saving the db to file in compressed format and you can decide on which way is better.

**Comment by harbaughfan** [comment]
Hi, I added a routine to load_database_from_file where the file is previously generated by mcp_database.  So there's 3 ways it loads the db now,
First (Preferred): try to load a previously saved db file generated by save_db_to_file.
Second: look for blocks.json & items.json and parse those into db. (Saving our own file afterwards).
Third:  download server.jar and invoke java to extract blocks.json & items.json.  Then proceed with second method.

**Comment by sunkisser** [comment]
Hi just an update, the database loader now looks for blocks_404.json and items_404.json in the ./database directory.  previously it was looking for just blocks.json and items.json.

The database.c is completely working for protocol 404, I've gotten past all the issues of which variables are defined where.  
I've begun the conversion to multi-protocol & now just going through it continuing to convert to full multi-protocol ability.

**Comment by harbaughfan** [comment]
Hi just an update, the database loader now looks for blocks_404.json and items_404.json in the ./database directory. previously it was looking for just blocks.json and items.json.

The database.c is completely working for protocol 404, I've gotten past all the issues of which variables are defined where.
I've begun the conversion to multi-protocol & now just going through it continuing to convert to full multi-protocol ability.

**Comment by harbaughfan** [comment]
Hi, I added unload_all_databases() so I think the database module is ready to go. 

I'm ready to work on something else, can you think of another project for me?  I have all day tomorrow to get a great start on it.

**Comment by broese** [comment]
Hi, sorry I did not respond this week - been down with the flu and could not work on it. Thanks for the database module, I merged it. Pretty cool stuff with automatic download and extraction. I had to fix the Cygwin/Linux differentiation - Linux doesn't have the "OS" variable, so it's probably better to just use the platform macros the compiler provides. Can you please check if it still works for you?
Right now I've been trying to check if the chunk data is collected properly in the gamestate and it doesn't seem to be the case. Looks OK in the packet decoder, but somehow it's not stored as intended.
I just saw some important data is now missing for the items - we need some table for stack sizes. Some items are 16- and some non-stackable and they did not include this info in the items.json. Without it, the inventory tracking is not working properly. Maybe if you could make a table mapping the item name to the stack size and provide a lookup function? Also some flag indicating what type of block it is - plain, stairs, slab, etc. and if it's an armor or edible item - basically the I_XXXX flags in the old item table. What else - I don't know, maybe SP_Map packet implementation? Sorry, I'm still a bit ill atm, I'll check tomorrow for more.

**Comment by harbaughfan** [comment]
Hi! the platform macros to determine OS worked great for me on Cygwin.  
I'll start with incorporating stack size info.

**Comment by broese** [comment]
I moved the database issue to a separate thread, so it doesn't mix with the general discussion.
I need a few more things there in order to proceed with the build function. Maybe we can get at least simple blocks working soon.

**Comment by harbaughfan** [comment]
Regarding the item stack size, the only programmatic way I could find to get this info was to scrape off gamepedia.  Gotta be a cleaner way so if you ever run across another source of stacksize info, please post here & I'll update.  For now to get the info, I resort to the python script below .  But I have it!  I'll incorporate the stacksizes into database_c soon.
`from bs4 import BeautifulSoup
import urllib.request
import time

def getstacksizefromwebpage(fh):
    soup = BeautifulSoup(fh, 'html.parser')
    tabletext = soup.find("table",{"class":"infobox-rows"}).get_text().split()
    stackable = tabletext.index("Stackable")+1
    if tabletext[stackable] == "Yes":
        stacksize = tabletext[stackable+1].strip("()")
    else:
        stacksize = 1
    return stacksize

def openpage(url):
    response = urllib.request.urlopen(url)
    html = response.read()
    return html

def getitemlistfrommainpage(html):
    soup = BeautifulSoup(html,'html.parser')
    h3s = soup.find_all("h3")
    for h3 in h3s:
        if h3.get_text() == "Items[edit]":
            break

    itemtable = h3.find_next("table")
    rows = itemtable.find_all("tr")
    data = []
    for row in rows:
        tds = row.find_all("td")
        if len(tds) > 1:
            linkhtml = tds[0].find("a").get("href")
            itemname = tds[1].get_text().strip().split("\u200c")[0]
            data.append(["https://minecraft.gamepedia.com"+linkhtml,itemname])
    return data

def main():
    html = openpage("https://minecraft.gamepedia.com/Java_Edition_data_values")
    data = getitemlistfrommainpage(html)
    outputfile = open("Output.txt", "w")
    for row in data:
        if row[1] == "pufferfish":
            row[0] = row[0]+"_(item)"
        html = openpage(row[0])
        outputfile.write(row[1] + " " + str( getstacksizefromwebpage(html)) +"\n")
        print(row[1])
    outputfile.close
    return

main()`

**Comment by harbaughfan** [comment]
I just found http://minecraft-data.prismarine.js.org/?v=1.13.2
which gives the stacksizes for every item, as well as a ton of other info.  I'll get stacksize data from that at some point in the future.

**Comment by harbaughfan** [comment]
Scrap the above link.. it contained faulty info.  So did gamepedia for that matter.  I used both sources side by side & went through checking against my own knowledge & verifying in-game when I was unsure.

I can't decide between a lookup table between name & stacksize (790 items), or a table between itemid & stacksize (more compact array but we need one for each protocol) 
Below is what I came up with that uses the ID way. It works, but I think it might get ugly over time, since it needs a complete extra table with every version.
So I think I need your judgement on this.  After seeing the below id=>stacksize code, do you think this way works well, or does a name=>stacksize table seem better?
`const int stacksize404arr[] = {64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,1,1,1,1,1,1,
1,1,1,1,1,1,1,1,1,1,1,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
64,64,64,64,64,1,64,1,1,1,1,64,1,64,64,64,64,64,64,1,1,1,1,1,1,1,1,1,1,1,1,1,64,64,1,1,1,1,1,
64,64,64,1,1,1,1,1,64,64,64,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,64,64,64,64,64,64,16,16,1,
1,1,1,64,16,1,64,1,1,1,1,1,64,64,64,64,64,64,64,64,1,1,16,64,1,64,64,64,64,64,64,64,64,64,64,
64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,64,64,1,64,
64,64,64,64,64,64,64,64,16,64,64,64,64,1,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
64,64,64,64,64,64,64,64,64,64,64,64,64,64,1,16,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,1,
64,64,64,64,1,64,64,1,1,64,64,64,64,1,64,64,16,1,1,1,64,64,1,64,64,16,16,16,16,16,16,16,16,16,
16,16,16,16,16,16,16,64,64,64,64,64,1,64,1,1,1,1,1,1,1,1,1,1,1,1,64,64,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,64,64,64};`
`int stacksize(database_t *db, char *itemname) {`
    `    int itemid = get_item_id(db, itemname);`
`    if ( itemid < 0 || itemid >= sizeof(stacksize404arr)/sizeof(int) ) {`
`        return 0;`
`    }`
`    return stacksize404arr[itemid];`
`}`

The alternate name=>stacksize way would be 790 lines of code, but easier to maintain with future versions. Assuming stack size never changes for an item between versions.

{ {"air",64},
{"stone",64},
{"granite",64},
{"polished_granite",64},
{"diorite",64},
{"polished_diorite",64},
{"andesite",64},
{"polished_andesite",64},
{"grass_block",64},
{"dirt",64},
{"coarse_dirt",64},
{"podzol",64},
{.  . .}, _(770+ lines of code)_
{"trident",1},
{"phantom_membrane",64},
{"nautilus_shell",64},
{"heart_of_the_sea",64} };

**Comment by broese** [comment]
I would suggest some very simple format for the table. Technically it can be just an array of uint32_t (or maybe rather uint64_t, it was getting a bit full), indexed by item_id and the values are OR-ed I_XXXX flags, just like they were used in the past. Stack size was defined by I_NSTACK and I_S16 flags.  Later I can add flags for various block types (e.g. I_STAIR, I_SLAB, etc.) manually.
I will need to lookup things for the blocks (i.e. already placed) as well, in that case I guess I will have convert block_id -> name -> item_id -> flags. Hmm,.. this lookup gets maybe a bit inefficient, but it should be fine for now. We can optimize the database organization later.
I really hope Mojang will just extend the list of items and not shuffle their ID. The exported lists look like that at least for 1.13.0..2 versions.

**Comment by broese** [comment]
So, maybe something like:
```
{
[2] = 0, // granite
...
[523] = I_NSTACK|I_ITEM|I_ARMOR, // iron_helmet
...
[624] = I_S16|I_ITEM, // enderpearl
...
}
```

**Comment by harbaughfan** [comment]
That looks perfect, I'll dive in tonight after work.

**Comment by harbaughfan** [comment]
One question, do you see a long term direction of mcp_ids.c/h going away, replaced by new functionality of mcp_database?   Or will the two modules still co-exist with different purposes.

**Comment by broese** [comment]
Yes, this would be very reasonable, since most of the things there are related to the old id/meta system handling. And protocol/packet IDs could be moved to mcp_packet.h. At the moment I have to leave a lot of legacy things as they are, avoiding breaking compilation in hundreds of places. I'd also like to split the inventory tracking into a separate module as it takes too much space in mcp_gamestate

**Comment by broese** [comment]
GREAT SUCCESS!
First floor worked! With material fetching from inventory, material specification, etc. And with an unintentional funny sound effect :) .
This obviously only works with simple blocks (dirt/cobble/etc.) but it's already half the battle.
(The other half will be ridiculously difficult I'm afraid)

I've modified and moved two functions into database module, then I noted a few things - I'll put them into #30 thread.

At some point I would like to start using a typedef uint16_t bid_t for the block IDs instead of int, so we have a well-defined type. gamestate may accumulate quite an amount of blocks, so it's important to have them as a compact 16-bit type.

**Comment by harbaughfan** [comment]
Congrats!!!!!!!  It worked for me too, the sound effect is very appropriate!
I would be happy to convert to bid_t.

**Comment by harbaughfan** [comment]
Hi, I'll work on the conversion to uint16_t for the block id today.  Since there's already a bid_t, I'll use a distinct name `typedef uint16_t blid_t` for now.  Once we are able to remove the old bid_t,  then we can rename the new to bid_t.  

PS I was tinkering around last night & got slabs and mapart (wool) working!  I rebased this morning so it won't affect our workflow, but we are getting so close!! I'm excited!
